activitysim/core/configuration/base.py

```diff
@@ -1,32 +1,116 @@
 from __future__ import annotations
 
-from typing import Any, Union  # noqa: F401
+from pathlib import Path
+from typing import Any, Literal, TypeVar, Union  # noqa: F401
 
-from activitysim.core import configuration
+from pydantic import BaseModel as PydanticBase
 
-try:
-    from pydantic import BaseModel as PydanticBase
-except ModuleNotFoundError:
+from activitysim.core import configuration
 
-    class PydanticBase:
-        pass
+PydanticReadableType = TypeVar("PydanticReadableType", bound="PydanticReadable")
 
 
 class PydanticReadable(PydanticBase):
```

Comment:
A general question about the design: when does the upper-level settings (like settings.yaml, network_los.yaml) and the component settings files being validated? Are they validated all together in the beginning when a state object is initialized, or are they validated seperately and only when they are used, like the auto ownership settings will only be validated when its read in during the auto ownership step? The pro of the former is to not have to wait till the end to find out e.g., trip scheduling setting is wrong.

Reply:
Validation happens when a settings file is read.  It can't happen sooner than that, we need to know the content to do the validation.  The design choice to load on demand is old and is not changed by this PR.  I've opened a feature request #784 to consider this change separately.